### PR TITLE
Prebuilds Page

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Project Overview

Showing branch details.

Screen Shot 2021-07-22 at 12 34 02

---
## Prebuilds

Listing all prebuilds of a project.

This is the first iteration on https://github.com/gitpod-io/gitpod/issues/4515

Screen Shot 2021-07-22 at 12 34 07

---
## Prebuild

To show all details on a prebuild. Currently just including the existing `<PrebuildLogs>` components. The plan is to restyle it afterwards.

Screen Shot 2021-07-23 at 14 09 03

### Known Issue

We've no build numbers (yet)

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4876#" title="Author self-approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev]
- ~~[components/licensor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/licensor/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4876#" title="Author self-approved">AlexTugarev</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [AlexTugarev]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev]
- ~~[components/licensor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/licensor/OWNERS)~~ [AlexTugarev]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (AlexTugarev) [inline components/dashboard/src/components/PrebuildLogs.tsx:18]: @jankeromnes logs were sent to the wrong emitter. `const` values should not be state. now I can see logs and log output for prebuilds 👍🏻

**Comment by jankeromnes** [comment]
When I try to add a new Project to a new Team, and I click on "Configure", I get some sort of error page. Is that expected?

<table><tr><td>
Screenshot 2021-07-23 at 14 22 32
</td><td>
Screenshot 2021-07-23 at 14 22 42
</td></tr></table>

**Comment by AlexTugarev** [comment]
@jankeromnesl, that's always the case when the GH App config is gone. Triggering werft again...

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (jankeromnes) [reply components/dashboard/src/components/PrebuildLogs.tsx:18]: Aha, exciting!

**Comment by jankeromnes** [comment]
Wow, very cool! 🙌 I was able to see a newly triggered prebuild (although it took quite a long time before showing up), and then check on its logs.

I love it 😍

Screenshot 2021-07-23 at 14 39 23
Screenshot 2021-07-23 at 14 39 35

**Thought**: When adding a new project, I'd love to see all the older prebuilds as well. Can't we also surface older prebuilds? (E.g. based on the clone URL).

File: components/dashboard/src/service/service.tsx
Comment (jankeromnes) [inline components/dashboard/src/service/service.tsx:57]: **Thought**: Cool! I realize that this trick could also be used by admins to debug user-specific bugs (e.g. a subscription problem where we don't understand how the Plans page can get into that state -- for these cases, maybe an admin could visit `/?userId=1234` and view the app as user 1234 -- with appropriate audit logging to cross-verify admin activity)

File: components/dashboard/src/workspaces/workspace-model.ts
Comment (jankeromnes) [inline components/dashboard/src/workspaces/workspace-model.ts:139]: **Nit**: Why do you consider deleted/garbage-collected workspaces as active?

File: components/gitpod-db/src/project-db.ts
Comment (jankeromnes) [inline components/gitpod-db/src/project-db.ts:14]: A second time? (See first method)

File: components/gitpod-db/src/project-db.ts
Comment (jankeromnes) [inline components/gitpod-db/src/project-db.ts:15]: **Nit**:

```suggestion
    findProjectByTeamAndName(teamId: string, projectName: string): Promise<Project | undefined>;
```

File: components/licensor/typescript/ee/src/index.ts
Comment (jankeromnes) [inline components/licensor/typescript/ee/src/index.ts:45]: **Reminder**: Please revert this commit before merging.

File: components/server/src/auth/resource-access.ts
Comment (jankeromnes) [inline components/server/src/auth/resource-access.ts:95]: **Nit**: I think the guarded resource is always the team. Please remove this for now (especially since it isn't used anywhere).

File: components/gitpod-protocol/src/gitpod-service.ts
Comment (jankeromnes) [inline components/gitpod-protocol/src/gitpod-service.ts:311]: **Nit**: As before, I still think these parameters do not make sense in `CreateWorkspaceOptions`, and also they aren't used anywhere in this PR. Please remove them for now. 😇

```suggestion
```

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (jankeromnes) [inline components/gitpod-protocol/src/teams-projects-protocol.ts:42]: **Nit**: This is the only field of `BranchDetails` that has the `branch` prefix. Maybe it's not needed?

```suggestion
        url: string;
```

**Review by jankeromnes — COMMENTED** [review]
This is amazing! 😁 Thanks again for powering through this and displaying prebuilds & their status/logs! 🙏

This is quite a big PR. I've done a quick pass on the code, and left some comments/suggestions/thoughts in-line.

Also, have a fantastic weekend! ☀️

**Comment by AlexTugarev** [comment]
@jankeromnes 

> It took quite a long time before my prebuild showed up (it did well after finishing). Do you know why there is a long delay between the trigger of a prebuild and it showing up? (Is that db-sync?)

Event handling for new prebuilds is not part of this PR. You'd need to reload the page manually to see changes. 

> When adding a new project, I'd love to see all the older prebuilds as well. Can't we also surface older prebuilds? (E.g. based on the clone URL).

Though technically possible, I'd argue to not pull in this complexity. Having a clear relation between a prebuilt workspace and a project makes the implementation cleaner then dealing with branch mismatches and the like. OTOH we can revisit this when migrating existing app installations, wdyt?

File: components/dashboard/src/components/PrebuildLogs.tsx
Comment (AlexTugarev) [reply components/dashboard/src/components/PrebuildLogs.tsx:19]: Looks a bit exceptional, but this is just to emphasize not to reuse the instance in async code or in subcomponents.

File: components/dashboard/src/App.tsx
Comment (AlexTugarev) [reply components/dashboard/src/App.tsx:210]: I tried it, it wasn't working as expected. Please have a look at it separately 🙏🏻

File: components/dashboard/src/workspaces/workspace-model.ts
Comment (AlexTugarev) [reply components/dashboard/src/workspaces/workspace-model.ts:139]: I'm not, but I'll revert anyways. 
Looking at the workspace-db-impl.ts it inconsistently checking for `softDeleted`.

File: components/server/src/auth/resource-access.ts
Comment (AlexTugarev) [reply components/server/src/auth/resource-access.ts:95]: removing...

File: components/gitpod-protocol/src/gitpod-service.ts
Comment (AlexTugarev) [reply components/gitpod-protocol/src/gitpod-service.ts:311]: will do...

File: components/dashboard/src/workspaces/workspace-model.ts
Comment (AlexTugarev) [reply components/dashboard/src/workspaces/workspace-model.ts:139]: reverting...

File: components/gitpod-db/src/project-db.ts
Comment (AlexTugarev) [reply components/gitpod-db/src/project-db.ts:14]: failed git merge. removing...

File: components/gitpod-protocol/src/teams-projects-protocol.ts
Comment (AlexTugarev) [reply components/gitpod-protocol/src/teams-projects-protocol.ts:42]: fair point.

File: components/dashboard/src/projects/Project.tsx
Comment (svenefftinge) [inline components/dashboard/src/projects/Project.tsx:194]: This button should allow CMD+Click

Comment (svenefftinge) [reply components/dashboard/src/projects/Project.tsx:194]: but can be done in follow up

File: components/dashboard/src/App.tsx
Comment (gtsiolis) [inline components/dashboard/src/App.tsx:39]: **issue**: Fetching active branches and prebuilds requires a noticable amount of time, 5 seconds or more. Is there anything we could do in this iteration to make this smoother? If not, let's add a follow up issue for adding a loading indicator in a future iteration. ⏱️

**Review by gtsiolis — COMMENTED** [review]
Great work @AlexTugarev! 🌟 🌟 🌟 

Feels so good to see prebuilds on the dashboard! 🧡 

Added a few comments below, let me know if anything sounds confusing and feel free to open follow up issues to address smaller issues.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (svenefftinge) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1476]: don't forget to put this in a list of todos

**Review by svenefftinge — APPROVED** [review]
LGTM. I think we can merge this and continue work in smaller steps

**Comment by roboquat** [comment]
LGTM label has been added.  <details>Git tree hash: c38192ab1a8f78d1522a3c24f99166225951bb1f</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/4876#" title="Author self-approved">AlexTugarev</a>*, *<a href="https://github.com/gitpod-io/gitpod/pull/4876#pullrequestreview-714908185" title="Approved">svenefftinge</a>*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/gitpod-io/gitpod/blob/main/OWNERS)~~ [AlexTugarev,svenefftinge]
- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [AlexTugarev,svenefftinge]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [AlexTugarev,svenefftinge]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [AlexTugarev,svenefftinge]
- ~~[components/licensor/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/licensor/OWNERS)~~ [AlexTugarev,svenefftinge]
- ~~[components/server/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/server/OWNERS)~~ [AlexTugarev,svenefftinge]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
